zerver/actions/message_edit.py

```diff
@@ -1363,3 +1365,43 @@ def check_update_message(
         queue_json_publish("embed_links", event_data)
 
     return number_changed
+
+
+@transaction.atomic
+def unmove_topic(
+    user_profile: UserProfile, stream_id: int, from_topic_name: str, to_topic_name: str
+) -> int:
+    stream = access_stream_by_id(user_profile, stream_id, require_active=True)[0]
+    recipient_id = assert_is_not_none(stream.recipient_id)
+
+    move_messages = (
+        Message.objects.filter(
+            realm_id=stream.realm_id,
+            recipient_id=recipient_id,
+            edit_history__isnull=False,
+            subject=from_topic_name,
+        )
+        .annotate(edit_history_json=Cast("edit_history", output_field=JSONField()))
+        .filter(
+            edit_history_json__0__prev_topic=to_topic_name,
+        )
+        .select_for_update()
+    )
+
+    message_count = 0
+
+    for message in move_messages:
+        message_count += do_update_message(
+            user_profile,
+            message,
+            new_stream=None,
+            topic_name=to_topic_name,
+            propagate_mode="change_one",
```

Comment:
This seems fine; though it does make me think we should consider making a `change_several` mode, perhaps one that isn't even exposed in the API, to avoid the loop here.

I guess it would also not be hard to support `new_stream` not being `None`.